components/engine_traits/src/errors.rs

```diff
@@ -1,55 +1,41 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
+use std::{error, result};
+
 use error_code::{self, ErrorCode, ErrorCodeExt};
 use raft::{Error as RaftError, StorageError};
-use std::{error, result};
+use thiserror::Error;
 
-quick_error! {
-    #[derive(Debug)]
-    pub enum Error {
-        // Engine uses plain string as the error.
-        Engine(msg: String) {
-            from()
-            display("Storage Engine {}", msg)
-        }
-        // FIXME: It should not know Region.
-        NotInRange( key: Vec<u8>, region_id: u64, start: Vec<u8>, end: Vec<u8>) {
-            display(
-                "Key {} is out of [region {}] [{}, {})",
-                log_wrappers::Value::key(&key), region_id, log_wrappers::Value::key(&start), log_wrappers::Value::key(&end)
-            )
-        }
-        Protobuf(err: protobuf::ProtobufError) {
-            from()
-            cause(err)
-            display("Protobuf {}", err)
-        }
-        Io(err: std::io::Error) {
-            from()
-            cause(err)
-            display("Io {}", err)
-        }
-        Other(err: Box<dyn error::Error + Sync + Send>) {
-            from()
-            cause(err.as_ref())
-            display("{:?}", err)
-        }
-        CFName(name: String) {
-            display("CF {} not found", name)
-        }
-        Codec(err: tikv_util::codec::Error) {
-            from()
-            cause(err)
-            display("Codec {}", err)
-        }
-        EntriesUnavailable {
-            from()
-            display("The entries of region is unavailable")
-        }
-        EntriesCompacted {
-            from()
-            display("The entries of region is compacted")
-        }
+#[derive(Debug, Error)]
+pub enum Error {
+    // Engine uses plain string as the error.
+    #[error("Storage Engine {0}")]
+    Engine(String),
+    // FIXME: It should not know Region.
+    #[error(
+        "Key {} is out of [region {}] [{}, {})",
+        log_wrappers::Value::key(.0), .1, log_wrappers::Value::key(&.2), log_wrappers::Value::key(&.3)
+    )]
+    NotInRange(Vec<u8>, u64, Vec<u8>, Vec<u8>),
+    #[error("Protobuf {0}")]
+    Protobuf(#[from] protobuf::ProtobufError),
+    #[error("Io {0}")]
+    Io(#[from] std::io::Error),
+    #[error("{0:?}")]
+    Other(#[from] Box<dyn error::Error + Sync + Send>),
+    #[error("CF {0} not found")]
+    CFName(String),
+    #[error("Codec {0}")]
+    Codec(#[from] tikv_util::codec::Error),
+    #[error("The entries of region is unavailable")]
+    EntriesUnavailable,
+    #[error("The entries of region is compacted")]
+    EntriesCompacted,
+}
+
+impl From<String> for Error {
+    fn from(err: String) -> Self {
+        Error::Engine(err)
```

Comment:
This is too broad to be implemented. How about removing the trait and using `Error::Engine` directly?

Reply:
There may be dozens of implicit conversions about this in `engine_rocks`, I am not sure whether to remove this trait.

Reply:
So terrible. 😓